### ISSUE TITLE
PyOpenMS / numpy memoryview ug

### ISSUE THREAD

**Issue by hroest** [OpenMS/OpenMS#5862]
When working with float data arrays I ran into a very strange bug where the underlying memory view would somehow behave strange. 

What seems to trigger the bug is when a FloatDataArray is accessed with `get_data` and subsequently new memory is allocated:

```
drift = spec.getFloatDataArrays()[0].get_data()
bugtrigger = numpy.random.rand(39438)

```

leading to corruption of the data held in `drift`. The following test file and python program demonstrates the issue.

[bug.zip](https://github.com/OpenMS/OpenMS/files/8145639/bug.zip)
 
Furthermore, it seems that only pyopenms==2.7.0 triggers the bug:
```
$ python triggerBug.py 
<module 'pyopenms' from '/home/hroest/.local/lib/python3.8/site-packages/pyopenms/__init__.py'>
PyopenMS Version 2.7.0
Numpy Version 1.22.2
98.20384101719029 8.0 -1.0309144e-25 1.0726626 39438 39438
Bug: -1.0309144e-25 != 1.0726626 
End of Test
```
and not earlier versions:

```
$ python triggerBug.py 
<module 'pyopenms' from '/home/hroest/.local/lib/python3.8/site-packages/pyopenms/__init__.py'>
PyopenMS Version 2.6.0
Numpy Version 1.22.2
End of Test
```
which is strange as the corresponding commit has been from 2018 091cfdcd8341b8693f537df6dd9656985ea

**Comment by jpfeuffer** [comment]
Hmm but there was a commit shortly after this one that switches to memory views on the underlying vector data.
I guess depending on how quick the GC is, your float data array is already deallocated once you reach the next line.

**Comment by jpfeuffer** [comment]
Try saving your float data array py object to a variable first

**Comment by mobiusklein** [comment]
The best way around this might be to implement the [buffer protocol](https://cython.readthedocs.io/en/latest/src/userguide/buffer.html#a-matrix-class), which does two things: allows you to export the underlying memory buffers directly, and manages the lifetime of the exporting object so that it lives as long as it is borrowed (the `Py_buffer` object holds a reference to the exporting object, and the `Py_buffer` object is held by the object that is consuming the exported buffer).

It also makes it so you can directly do `np.asarray(openms_float_array)` without a copy from Python to get the same behavior you get from the `get_data()` method. You could then simply do `np.asarray(self)` in `get_data` for compatibility.

I've looked at the autowrap code many times with varying degrees of comprehension, but it seems like it should be possible to add another directive that adds the required code (and attributes) automatically for a specified attribute, though I can't tell whether the code generator knows the "type" of an attribute, so it might be a bit verbose.

**Comment by mobiusklein** [comment]
I've opened a draft PR https://github.com/OpenMS/autowrap/pull/126 to add buffer protocol support there.

**Comment by jpfeuffer** [comment]
While your PR is a great addition that we were discussing some time ago already for autowrap (to have a choice between views and copies), I am not sure if this completely solves the problem here. If we leave the memory management to python with the buffer protocol, we need to make sure that the C++ object (in this case MSSpectrum), does not deallocate the data in the background once it is deleted. For this we only have three choices:
- make a copy
- set the buffer in the C++ object to some safe place (making it non-sense inside the C++ object and its Python wrapper; basically a swap)
- modify the destructor to only deallocate if there is no python buffer pointing to it (basically do an own reference counting)

**Comment by jpfeuffer** [comment]
We are actually happy about some feedback here. We restrained from implementing too many "views", yet, since beginner python programmers might not be very familiar with such concepts and this might leave them with hard-to-debug errors.

**Comment by hroest** [comment]
@jpfeuffer the very easy solution to this, and in line with your 3rd point, would be to share ownership between Python and C++ which would be close to your 3rd point. Here the solution would be to use `shared_ptr` to pass between C++ and Python, allowing objects created on the C++ side to be passed over to Python and the other way around.

**Comment by mobiusklein** [comment]
Isn't the autowrap-generated `FloatDataArray` (the generated Python wrapper) really just a wrapper around a `shared_ptr` to the C++ `FloatDataArray` object which owns the `std::vector<float>` which owns the block of memory being read? The whole point here is that the borrowed memory is deallocated while the Python object using it is still alive.

It's possible that a change on C++ side can reallocate that `vector` and cause the data to move somewhere else in memory, but that'd be true for any operation that takes a reference to the underlying raw pointer the `vector` owns. Essentially adding a read-write-lock to each array is the only way to do this "safely", and that's not ideal on many levels.

On the Python side, we can make the buffer read-only which prevents Python code from mutating/reallocating the referenced memory. The user has to know that `np.array(x)` makes a copy unless `copy=False` is passed, and that `np.asarray(x)` borrows the buffer from `x` if it has one, otherwise does the normal `PySequence_GetItem` loop to copy data out into a freshly created `np.ndarray`. 

It's true if you have two methods, `get_data` and `get_view`, one explicitly copying and the other borrowing, you minimize the potential for bugs, but doesn't it still leave existing code vulnerable to the memory corruption problem discussed here?

**Comment by hroest** [comment]
@mobiusklein yes that is essentially true, FloatDataArray is a wrapper around a shared ptr. However the issue is not that the vector gets changed, the issue here is that the FloatDataArray *itself* is only temporary: `spec.getFloatDataArrays()` will return a temporary copy of float data arrays and `get_data` gets a memory view so doing 

```
drift = spec.getFloatDataArrays()[0].get_data()
```

is equivalent to

```
def getView(spec):
  tmp_copy = spec.getFloatDataArrays()
  resulting_view = tmp_copy[0].get_data()
  return resulting_view

```
where a temporary copy of all float data arrays is made and then on that *temporary* copy a view is created and returned. Now, as soon as the temporary copy goes out of scope if can be garbage collected and therefore the view would become invalid.

I think in the end, the only real "safe" thing to do here is to make a copy and not return the view, especially for float data arrays where this is easily missed

**Comment by mobiusklein** [comment]
Thank you for unpacking the problem, and tolerating my comments. If `FloatDataArray` C++ object doesn't own its data,  I do not understand which object actually owns the lifetime of the `vector<float>` then, and I'll stop interjecting.

This was my understanding of the current memory ownership relationship:
```mermaid
flowchart TD;
    view_of_float*:::python -. borrowing .->float*;
    Py-FloatDataArray:::python -- owning --->shared_ptr_FloatDataArray["shared_ptr&lt;FloatDataArray&gt;"];
    shared_ptr_FloatDataArray -- owning --->FloatDataArray;
    FloatDataArray-- owning --->vector_float["vector&lt;float&gt;"];
    vector_float-- owning --->float*

    classDef python fill:blue, stroke:yellow, stroke-width:2px
```
This is what the my proposal would cause:
```mermaid
flowchart TD;
    view_of_float*:::python -. borrowing .->float*;
    view_of_float*-- owning --->Py-FloatDataArray;
    Py-FloatDataArray:::python -- owning --->shared_ptr_FloatDataArray["shared_ptr&lt;FloatDataArray&gt;"];
    shared_ptr_FloatDataArray -- owning --->FloatDataArray;
    FloatDataArray-- owning --->vector_float["vector&lt;float&gt;"];
    vector_float-- owning --->float*

    classDef python fill:blue, stroke:yellow, stroke-width:2px
```

**Comment by jpfeuffer** [comment]
Hmm I think the problem might be, that your PyFloatDataArray does not own shared_ptr<FloatDataArray> but shared_ptr<FloatDataArray>.data.
This does not increase the reference count to the parent object. If the original FloatDataArray goes out of scope it will deallocate the whole object, including data, which then results in nonsense values for your PyFloatDataArray.

**Comment by mobiusklein** [comment]
Returning to the Python code,
```python
data_array = spec.getFloatDataArrays()[0] <-- creates `PyFloatDataArray` wrapping the C++ object
resulting_view = data_array.get_data() <-- creates a view of the C++ object's memory, `view_of_float*`
```
If keeping `data_array` alive prevents the memory `resulting_view` points to from being deallocated, then `PyFloatDataArray` (the autowrap generated class that wraps the C++ object) has an owning relationship with that memory. That was the solution that we all converged on earlier, correct?

If that's true, then anything that keeps an owning reference to `data_array` will prevent this issue, which is what the second ownership graph would do.

**Comment by jpfeuffer** [comment]
But doesn't your view only keep data_array.data "alive"? You would need to keep a shared pointer to the whole FloatDataArray object to increase its reference count and prevent it from deallocating.

**Comment by mobiusklein** [comment]
Using the buffer protocol as I outlined earlier would keep the whole Python object (`data_array` the variable, of type `pyopenms.FloatDataArray` which I was referring to as `PyFloatDataArray` ambiguously) wrapping the C++ `FloatDataArray` object alive.

When you implement the buffer protocol directly on the type, not just wrapping a raw pointer in a buffer using Cython's convenient syntax,  you can tell the `Py_buffer` object that is used to transport the buffer from one object to another to also keep a reference to the exporting object. This way you can ensure that the lifespan of the borrowed memory lasts as long as the borrow does if the borrowed memory is tied to the lifespan of the exporting object, or to perform some operation whenever the buffer is borrowed and/or when a borrow ends.

I set this to occur automatically in `autowrap`: https://github.com/OpenMS/autowrap/pull/126/files#diff-58e77d1a56385ce4d178af5cf468bab830acb492e5cfc22d7d8e258abf5c232eR556.

This works on the assumption that the object being wrapped "owns" its data. If it is just a borrowed view of some other data, then it's just borrowing trouble unless something guarantees that owning object outlives its borrowers.

**Comment by jpfeuffer** [comment]
Ah so you want to change how pyopenms.FloatDataArray works internally, not what pyopenms.FloatDataArray.get_data returns?
Yes should work.

**Comment by mobiusklein** [comment]
Yes, that is what I propose. In principle, it could have been an addon fed to `autowrap`, like where `get_data` is implemented, that was hand written for a small number of classes, but considering how much repeated boilerplate code would be involved I thought it made more sense to put in `autowrap` itself.

**Comment by jcharkow** [comment]
was this issue addressed? Ran into this issue and it is quite annoying!
